### DISCUSSION TITLE
Washed-out colors for PNG output

### DISCUSSION THREAD

**Discussion by ghost** [ImageMagick/ImageMagick#4208] [Help]
### ImageMagick version

7.1.0-4

### Operating system

Arch Linux

### Description

PNG files produced by ImageMagick have washed-out colors.

![screenshot](https://user-images.githubusercontent.com/68487118/130351272-09faa0ae-3a7b-4c85-92fa-29d50293f925.png)

The following steps will:
1. Upsample `input.png` 8x so only the DC coefficient is used in JPG compression
2. Downscale the JPG to get `decoded.png`

This is a proof of concept that there is a lossless way with JPG (credits to @jonsneyers).

The output should be the same as input, although it seems that ImageMagick's `convert` produces PNG with weird gamma or washed-out colors.

However, Firefox and Chrome display the PNG with expected colors, but for Gwenview the colors are washed-out.

Another user has reported a similar issue in KDE bug tracker (https://bugs.kde.org/show_bug.cgi?id=441162). I would like to know whether this is an issue in ImageMagick or in Gwenview.

### Steps to Reproduce

1. `convert input.png -sample 800% upsampled.ppm; cjpeg -quality 100 -rgb -sample 1x1 upsampled.ppm > foo.jpg`
2. `convert foo.jpg -sample 12.5% decoded.png`

### Images

[Test images](https://github.com/ImageMagick/ImageMagick/files/7027088/images.zip)

**Comment by snibgo** [comment]
In Windows Photo Viewer and Gimp, input.png and decoded.png look the same.

So, the only viewer that shows washed-out colours is Gwenview. Is that correct?

If so, the bug would seem to be in Gwenview.

I suggest you reduce the command sequence to the minimum that shows the problem. 

The ppm format cannot record metadata such as "exif:ColorSpace", so this is removed. Perhaps Gwenview relies on that metadata.  Perhaps inserting that metadata with exiftool fixes the problem.

**Comment by ghost** [comment]
The `convert` command mentioned in the KDE bug can also be used to reproduce this issue.

For example, download [this image](https://www.lapisrelights.com/assets/img/anime/kv_img.png) and run:
`convert kv_img.png -background white -alpha remove kv_img-converted.png`

`kv_img.png` does not have any EXIF data, and it seems to be a plain sRGB PNG with transparency.

When I use `cjxl` to convert the test images and output images losslessly to JXL, the input PNGs -> JXLs output:
```
  red primaries XY: 0.639999 0.330010
  green primaries XY: 0.300004 0.600003
  blue primaries XY: 0.150002 0.059997
  transfer_function: 13
  rendering_intent: 1
```
While the problematic PNGs -> JXLs output:
```
  red primaries XY: 0.639999 0.330010
  green primaries XY: 0.300004 0.600003
  blue primaries XY: 0.150002 0.059997
  transfer_function: 65535
  transfer_function gamma: 0.454550
  rendering_intent: 1
```
(The above information is obtained through the `jxlinfo` tool in [libjxl (JPEG XL image reference implementation) repo](https://github.com/libjxl/libjxl))

The values of `transfer_function` and `transfer_function gamma` are weird for problematic PNGs.

**Comment by snibgo** [comment]
As you use IM v7, I suggest you use "magick" instead of "convert". I doubt that this will make a difference to your problem.

Please check that you are running v7.1.0-4. What does "convert -version" say?

The screenshot shown in the OP seems consistent with converting from linear RGB to non-linear sRGB twice. That conversion is approximately a power function. So we can do this twice in IM:
```
magick input.png -evaluate Pow 0.454545 -evaluate Pow 0.454545 x.png
```
x.png is:

![x](https://user-images.githubusercontent.com/33812368/130358729-44135696-3e5f-44f8-b944-f49f8c61bab7.png)

This suggests that Gwenview is confused about linear RGB versus non-linear sRGB.

**Comment by ghost** [comment]
> As you use IM v7, I suggest you use "magick" instead of "convert". I doubt that this will make a difference to your problem.

Unfortunately, replacing `convert` with `magick` in the commands makes no difference to this problem.

> Please check that you are running v7.1.0-4. What does "convert -version" say?

I swear that I am really using v7.1.0-4. There is no reason to fake the ImageMagick version I am using, and I have checked the version before submitting the issue:
`Version: ImageMagick 7.1.0-4 Q16 x86_64 2021-07-18 https://imagemagick.org`

**Comment by fmw42** [comment]
I see no evidence of that issue on IM 6.9.12.19 Q16 Mac OSX Sierra.

```
convert input.png -sample 800% upsampled.ppm
cjpeg -quality 100 -rgb -sample 1x1 upsampled.ppm > foo.jpg
convert foo.jpg -sample 12.5% decoded.png
compare -metric rmse input.png decoded.png null:
0 (0)
```

I suspect a lack of an sRGB profile and/or your viewer is causing the issue.
